### PR TITLE
#21: use binary search to make things faster in populate_component_ma…

### PR DESCRIPTION
**Pull request by dimatr**
…trix

### CODE REVIEW

**Comment by dimatr** [comment]
Things should be better with binary search

**Comment by josiahseaman** [comment]
I'm a big fan of binary search.  Is this PR ready for review?

File: segmentation.py [hunk 1]
```diff
@@ -38,11 +38,15 @@ def populate_component_occupancy(schematic: PangenomeSchematic):
...
 def populate_component_matrix(paths: List[Path], schematic: PangenomeSchematic):
-    for component in schematic.components:
-        # paths paths are in the same order as schematic.path_names
-        for i, path in enumerate(paths):
-            relevant = [bin for bin in path.bins if
-                        component.first_bin <= bin.bin_id <= component.last_bin]  # very costly loop
+    # the loops are 1) paths, and then 2) schematic.components
+    # paths are in the same order as schematic.path_names
+    for i, path in enumerate(paths):
+        sortedBins = SortedDict((bin.bin_id, bin) for bin in path.bins)
```
Comment (lomereiter) [inline segmentation.py]: Nitpick: please use snake_case like in the rest of the codebase

**Comment by dimatr** [comment]
yes, please.

File: segmentation.py [hunk 2]
```diff
@@ -11,7 +11,7 @@
 from pathlib import Path as osPath, PurePath
 from nested_dict import nested_dict
 from datetime import datetime
-
+from sortedcontainers import SortedDict
```
Comment (lomereiter) [inline segmentation.py:14]: This is a nice find, the library looks really solid!

File: segmentation.py (hunk 1 above)
Comment (dimatr) [reply segmentation.py]: got it, will do. It's a habit from the Java world

**Comment by subwaystation** [comment]
This looks really good, thank you @dimatr.
I checked the MD5 sums of an example run output of `master` and your `branch` and they match perfectly.
I can also report a speed improvement of at least 13x fold.

**Pull request merged**